wp-web-push/lib/delapuente/wp-sw-concoct/class-wp-sw-concoct.php

```diff
@@ -0,0 +1,111 @@
+require_once(__DIR__ . '/class-wp-sw-concoct-service-worker.php');
+
+class WP_SW_Concoct {
+    const SW_REGISTRAR_SCRIPT = 'wp-sw-concoct-manager';
+
+    private static $instance;
+
+    public static function init() {
+        if (!self::$instance) {
+            self::$instance = new self();
+        }
+    }
+
+    public static function get_manager() {
+        self::init();
+        return self::$instance;
+    }
+
+    private $router;
+
+    private $sw_registrar_script_url;
+
+    private $service_workers;
+
+    private $enabled_sw;
+
+    private function __construct() {
+        $this->router = WP_SW_Concoct_Router::get_router();
+        $this->enabled_sw = array();
+        $this->service_workers = array();
+        $this->setup_sw_registrar_script();
+    }
+
+    public function sw($scope='') {
+        if (!$scope) { $scope = $this->default_scope(); }
+        if (!array_key_exists($scope, $this->service_workers)) {
+            $this->add_new_sw($scope);
+        }
+        return $this->service_workers[$scope];
+    }
+
+    public function sw_js_id($scope='') {
+        if (!$scope) { $scope = $this->default_scope(); }
+        return $scope;
+    }
+
+    public function enable_sw($scope='') {
+        if (!$scope) { $scope = $this->default_scope(); }
+        $this->enabled_sw[] = $scope;
+    }
+
+    public function sw_registrar() {
+        header('Content-Type: application/javascript');
+        $contents = $contents = file_get_contents(__DIR__ . '/lib/js/sw-registrar.js');
+        $contents = str_replace('$enabledSw', $this->json_for_sw_registrations(), $contents);
+        echo $contents;
+        $this->end();
+    }
+
+    public function write_sw($scope) {
+        $service_worker = $this->service_workers[$scope];
+        header('Content-Type: application/javascript');
+        header('Service-Worker-Allowed: ' . $scope);
+        $service_worker->write_content();
+        $this->end();
+    }
+
+    private function default_scope() {
+        return home_url('/', 'relative');
+    }
+
+    private function add_new_sw($scope) {
+        $url = "wpswconcoct/sw/sw@$scope";
+        $route = $this->router->route($url);
+        $service_worker = new WP_SW_Concoct_Service_Worker($route);
+        $this->service_workers[$scope] = $service_worker;
+        $this->router->add_route($url, function () use ($scope) {
+            $this->write_sw($scope);
+        });
+    }
+
+    private function setup_sw_registrar_script() {
+        $this->sw_registrar_script_url = 'wpswconcoct/sw-registrar.js';
+        $this->router->add_route($this->sw_registrar_script_url,  array($this, 'sw_registrar'));
+        add_action('wp_enqueue_scripts', array($this, 'enqueue_registrar'));
+    }
+
+    public function enqueue_registrar() {
+        wp_register_script(self::SW_REGISTRAR_SCRIPT, $this->router->route($this->sw_registrar_script_url));
+    }
+
+    // As it is impossible to mock up exit, we abstract it into this end() method.
```

Comment:
`die` is another language construct (like `exit`) and `wp_die` formats the error with html so I prefer to wrap it anyway but thanks.